lib/Parser/Parser.yy

```diff
@@ -602,26 +780,26 @@ array_type:
     ;
 
 primary_type:
-      lParen type rParen
-    | lParen rParen arrow primary_type
-    | lParen type rParen arrow primary_type
-    | lParen function_type_param_types rParen arrow primary_type
+      lParen type rParen { $$ = $2; }
+    | lParen rParen arrow primary_type { $$ = $4; }
+    | lParen type rParen arrow primary_type { $$ = $5; }
+    | lParen function_type_param_types rParen arrow primary_type { $$ = $5; }
     | namespaced_identifier template_arguments_opt
```

Comment:
a type must be a `Ty`, and cannot be a RefExpr (should be an UnresolvedNameTy here)